src/grpc/projects.rs

```diff
@@ -105,328 +109,188 @@ impl ProjectService for ProjectServiceImpl {
-        //     log::error!("{}", e);
-        //     tonic::Status::aborted("Database request failed.")
-        // })?);
-        // Ok(tonic::Response::new(UpdateProjectDataClassResponse {
-        //     project,
-        // }))
-    }
-    async fn get_projects(
-        &self,
-        _request: Request<GetProjectsRequest>,
-    ) -> Result<Response<GetProjectsResponse>> {
-        todo!()
+        log_received!(&request);
+
+        let token = tonic_auth!(
+            get_token_from_md(request.metadata()),
+            "Token authentication error."
+        );
+
+        let request = DataClassUpdate::Project(request.into_inner());
+        let collection_id = tonic_invalid!(request.get_id(), "Invalid project id");
+        let ctx = Context::res_ctx(collection_id, DbPermissionLevel::WRITE, true);
+
+        tonic_auth!(
+            self.authorizer.check_permissions(&token, vec![ctx]).await,
+            "Unauthorized"
+        );
+
+        let project = tonic_internal!(
+            self.database_handler.update_dataclass(request).await,
+            "Internal database error."
+        );
+        self.cache
+            .update_object(&project.object.id, project.clone());
+        let project: generic_resource::Resource =
+            tonic_internal!(project.try_into(), "Collection conversion error");
+        let response = UpdateProjectDataClassResponse {
+            project: Some(project.into_inner()?),
+        };
+        return_with_log!(response);
     }
     async fn archive_project(
         &self,
-        _request: Request<ArchiveProjectRequest>,
+        request: Request<ArchiveProjectRequest>,
     ) -> Result<Response<ArchiveProjectResponse>> {
-        todo!()
-    }
-    async fn delete_project(
-        &self,
-        _request: Request<DeleteProjectRequest>,
-    ) -> Result<Response<DeleteProjectResponse>> {
-        todo!()
+        log_received!(&request);
+
+        let token = tonic_auth!(
+            get_token_from_md(request.metadata()),
+            "Token authentication error."
+        );
+
+        let request = SnapshotRequest::Project(request.into_inner());
+        let project_id = tonic_invalid!(request.get_id(), "Invalid dataset id.");
+        let ctx = Context::res_ctx(project_id, DbPermissionLevel::ADMIN, true);
+
+        tonic_auth!(
+            self.authorizer.check_permissions(&token, vec![ctx]).await,
+            "Unauthorized"
+        );
+
+        // this only contains one entry with a dataset
+        let resources = tonic_internal!(
+            self.database_handler.snapshot(request).await,
+            "Internal database error."
+        );
+        for resource in &resources {
+            self.cache
+                .update_object(&resource.object.id, resource.clone());
+        }
+        let project: generic_resource::Resource =
+            // First entry is always the archived project 
+            tonic_internal!(resources[0].clone().try_into(), "Dataset conversion error");
```

Comment:
Copy & paste only partially adapted.

```suggestion
        let project_id = tonic_invalid!(request.get_id(), "Invalid project id");
        let ctx = Context::res_ctx(project_id, DbPermissionLevel::ADMIN, true);

        tonic_auth!(
            self.authorizer.check_permissions(&token, vec![ctx]).await,
            "Unauthorized"
        );

        // This contains all the resources the snapshot includes below the project
        let resources = tonic_internal!(
            self.database_handler.snapshot(request).await,
            "Internal database error."
        );
        for resource in &resources {
            self.cache
                .update_object(&resource.object.id, resource.clone());
        }
        let project: generic_resource::Resource =
            // First entry is always the archived project 
            tonic_internal!(resources[0].clone().try_into(), "Project conversion error");
```